Probe the new ignore functionality and add standardised tests

prockenschaub opened the pull request:
### Scope
This pull requesst aims to test the recently added functionality to ignore subsets of the data when fitting the imputation models (see #32) and implement standardised tests via `testthat`. It also aims to update the help pages where this hasn't already been done by @stefvanbuuren 

**Note:** This PR does not include additional features discussed in #32, such as `filter.mids`, plots, or vignettes. I will create separate PRs for those later this week.

### Testing
- [x] Create tests for ignore in `mice`

- [x] Create tests for ignore in `mice.mids`

- [x] Create tests for ignore in `rbind.mids` and `rbind.mids.mids`

- [x] Update help pages

prockenschaub commented:
The way I went about testing whether `ignore` is honoured in `mice` is by implementing test cases with outliers for `pmm` and `norm`. I don't have a feeling for the implementation differences between univariate methods yet. Adding separate tests for each method is likely overkill. Would a simple test case suffice (e.g. pmm)? Should we include one test case for each `defaultMethod`?

stefvanbuuren commented:
Thanks for the work, and for correcting my glitches. 

I like your tests for mice ignore parameter. There was one failure, which can be solved by adding the `matchtype = 0` argument, so

```r
imp_pmm1 <- mice(
  rbind(nhanes, add_pmm),
  maxit = 1, m = 5, print = FALSE, seed = 1, donors = 1L, matchtype = 0,
)
```

and then tests run fine. I think testing pmm is enough. Other univariate methods just work in the same way.

stefvanbuuren commented:
One request: I would like to promote the `ignore` parameter to a more prominent location in the argument list, just before `where`. Could you make that move (also in the documentation)?

Why? Gerko Vink suggested to me a more general use of the `ignore` vector. At some later point, I wish to generalise `ignore` to a matrix of logical, which can take over the role of `ry`. This enables things like fitting the imputation model on the imputed values (this sounds circular but could be useful for MNAR scenarios), removing rows with skip patterns from the imputation model, and perhaps other use cases we don't know yet.

prockenschaub commented:
- I have add a number of further tests for `mice.mids` and `rbind.mids`

- I moved `ignore` in front of `where`  in the code and documentation of `mice`. I also moved the argument in front of mice for `initialize.imp` and `sampler` for consitency. These functions are sometimes called with just positional arguments, so I changed those calls too. I hope I haven't missed any call, the unit tests finish without error. 

- I have *not* moved the position of `ignore` in the `mids` object. Let me know if you would like to move that too. 

- While testing the `newdata` parameter in `mice.mids`, I found that I the behaviour of `mice.mids(imp, newdata = nhanes, maxit = 0)` ran counter to my intution, because it returned only imp. This is due to an if-clause at the very beginning that checks for `maxit < 1`. This made sense pre-`newdata` but I think should be changed now. I moved the appending of newdata before this if-clause.

- I removed checks for collinearity or constant variables for newdata. I don't think they make sense, since collinearity does not matter in those contexts (newdata isn't used to build the imputation model).

prockenschaub commented:
I am happy to give a more "general" version of `ignore` a go and can see if I can replace `ry` with it (separate PR). 

Would this change our rationale for `filter.mids`? I guess not, because this other way of using ignore would probably mostly be used from `mice` and not from `mice.mids(., newdata = x)`. If you agree witht that, I will create a new PR to look at `filter.mids`, plots, and vignettes.

stefvanbuuren reviewed (approved):
A well-prepared and thorough pull request. Didn't see any oversights, so I'm happy to merge.

stefvanbuuren commented:
Thanks, looking forward to any initiatives for matrix `ignore` and `filter.mids()`. 

I think that a `filter.mids()` should only accept a vector-like argument to subset the `mids` object. That could be a vector `ignore`. I would think that `filter.mids(mice.mids(., newdata = x), ignore)` should return the `mids` object pertaining to the subset of rows corresponding to `newdata` (of course, assuming here that all cases in the training `imp` have `ignore == FALSE`).

A side note: The usefulness of `filter.mids` increases if it could also do arbitrary selection based on the imputed data. But at the same time the prospect of having different rows in different imputed data sets worries me as the `mids` object was never designed to support such a structure, so perhaps we should forget about this option. The user might be better off using `tidyverse` tricks applied to `complete(imp, "long")`), although repeated analysis by `with.mids` and pooling by `pool()` is not possible.
